program/shell/errors_to_json.go

```diff
@@ -0,0 +1,47 @@
+package shell
+
+import (
+	"encoding/json"
+
+	"github.com/elves/elvish/eval"
+	"github.com/elves/elvish/parse"
+)
+
+// ErrorInJSON is the data structure for self-defined error
+type ErrorInJSON struct {
+	FileName string `json:"fileName"`
+	Start    int    `json:"start"`
+	End      int    `json:"end"`
+	Message  string `json:"message"`
+}
+
+// SimplifiedErrorInJSON is the data structure for errors which contain limited info
+type SimplifiedErrorInJSON struct {
+	Message string `json:"message"`
+}
+
+// ErrorToJSON converts the error into JSON format
+func ErrorToJSON(err error) string {
+	var e interface{}
+	switch err := err.(type) {
+	case *eval.CompilationError:
+		e = []interface{}{ErrorInJSON{err.Context.Name, err.Context.Begin, err.Context.End, err.Message}}
+	case *parse.MultiError:
+		e = processMultiError(err)
```

Comment:
You can inline the implementation of `processMultiError` here.

Reply:
I am definitely not the biggest fan of inlining all "simple" functions. Even though `processMultiError` is not called elsewhere, the function name could still help improve the readability.

Anyway here is just a loop, will do if you ask so